### ISSUE TITLE
Insert timeout

### ISSUE THREAD

**Issue by filip-halt** [milvus-io/milvus#7015]
Unable to insert past 5 million vectors. The connection times out and nothing can be inserted after. Once it fails the collections get buggy (https://github.com/milvus-io/milvus/issues/7013). Upon examination it seems like it is a pulsar error and I have attached its log here. 
```
20:34:55.484 [bookkeeper-ml-workers-OrderedExecutor-1-0] WARN  org.apache.bookkeeper.client.RackawareEnsemblePlacementPolicyImpl - Failed to find 1 bookies : excludeBookies [], allBookies [].
20:34:55.484 [bookkeeper-ml-workers-OrderedExecutor-1-0] WARN  org.apache.bookkeeper.client.RackawareEnsemblePlacementPolicyImpl - Failed to find 1 bookies : excludeBookies [], allBookies [].
20:34:55.484 [bookkeeper-ml-workers-OrderedExecutor-1-0] ERROR org.apache.bookkeeper.client.LedgerCreateOp - Not enough bookies to create ledger
20:34:55.484 [bookkeeper-ml-workers-OrderedExecutor-1-0] ERROR org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl - [public/default/persistent/query-node-stats] Error creating ledger rc=-6 Not enough non-faulty bookies available
20:34:55.484 [bookkeeper-ml-workers-OrderedExecutor-1-0] INFO  org.apache.pulsar.broker.service.Producer - Disconnecting producer: Producer{topic=PersistentTopic{topic=persistent://public/default/query-node-stats}, client=/10.0.2.96:46836, producerName=standalone-0-1, producerId=1}
20:34:55.484 [bookkeeper-ml-workers-OrderedExecutor-1-0] WARN  org.apache.pulsar.broker.service.persistent.PersistentTopic - [persistent://public/default/query-node-stats] Failed to persist msg in store: Not enough non-faulty bookies available
20:34:55.485 [pulsar-io-51-2] INFO  org.apache.pulsar.broker.service.persistent.PersistentTopic - [persistent://public/default/query-node-stats] Un-fencing topic...
20:34:55.485 [pulsar-io-51-2] INFO  org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl - [public/default/persistent/query-node-stats] Managed ledger is now ready to accept writes again
20:34:55.520 [pulsar-io-51-2] INFO  org.apache.pulsar.broker.service.ServerCnx - [/10.0.3.5:37954][persistent://public/default/proxyTimeTick-0] Creating producer. producerId=1
20:34:55.520 [Thread-15031] INFO  org.apache.pulsar.broker.service.ServerCnx - [/10.0.3.5:37954] persistent://public/default/proxyTimeTick-0 configured with schema false
20:34:55.520 [Thread-15031] INFO  org.apache.pulsar.broker.service.ServerCnx - [/10.0.3.5:37954] Created new producer: Producer{topic=PersistentTopic{topic=persistent://public/default/proxyTimeTick-0}, client=/10.0.3.5:37954, producerName=standalone-0-2, producerId=1}
20:34:55.522 [bookkeeper-ml-workers-OrderedExecutor-3-0] INFO  org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl - [public/default/persistent/proxyTimeTick-0] Creating a new ledger
20:34:55.522 [bookkeeper-ml-workers-OrderedExecutor-3-0] INFO  org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl - [public/default/persistent/proxyTimeTick-0] Creating ledger, metadata: {component=[109, 97, 110, 97, 103, 101, 100, 45, 108, 101, 100, 103, 101, 114], pulsar/managed-ledger=[112, 117, 98, 108, 105, 99, 47, 100, 101, 102, 97, 117, 108, 116, 47, 112, 101, 114, 115, 105, 115, 116, 101, 110, 116, 47, 112, 114, 111, 120, 121, 84, 105, 109, 101, 84, 105, 99, 107, 45, 48], application=[112, 117, 108, 115, 97, 114]} - metadata ops timeout : 60 seconds
20:34:55.522 [bookkeeper-ml-workers-OrderedExecutor-3-0] WARN  org.apache.bookkeeper.client.RackawareEnsemblePlacementPolicyImpl - Failed to find 1 bookies : excludeBookies [], allBookies [].
20:34:55.522 [bookkeeper-ml-workers-OrderedExecutor-3-0] WARN  org.apache.bookkeeper.client.RackawareEnsemblePlacementPolicyImpl - Failed to find 1 bookies : excludeBookies [], allBookies [].
20:34:55.522 [bookkeeper-ml-workers-OrderedExecutor-3-0] ERROR org.apache.bookkeeper.client.LedgerCreateOp - Not enough bookies to create ledger
20:34:55.522 [bookkeeper-ml-workers-OrderedExecutor-3-0] ERROR org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl - [public/default/persistent/proxyTimeTick-0] Error creating ledger rc=-6 Not enough non-faulty bookies available
```
#### Steps/Code to reproduce:
1. Run cluster on EKS
2. attempt to insert vectors in 100k batches of 128-dim vectors

#### Expected result:

Be able to fully insert

#### Actual results:

Fails to insert past 5 Million vectors. 

#### Environment:
v2.0.0-RC2 or 8b23a93
cluster
pymilvus v2.0.0rc2
AWS EKS

#### Configuration file:
helm upgrade --install --set dataNode.replicas=2 --set cluster.enabled=true --set service.type=LoadBalancer my-release milvus/milvus

**Comment by yhmo** [comment]
The error is thrown from Bookkeeper, a distribute storage system. The Pulsar use it to store data. The milvus start a Pulsar single instance, with Bookkeeper running on same container. When disk is fulled, the Bookkeeper thrown this error.
It is a third-party problem, so we suggest:
- check the pulsar container disk size
- specify Pulsar use secondary bookkeeper service when it doesn't have enough bookkeeper server available. https://pulsar.apache.org/docs/en/2.3.2/reference-configuration/.  The "bookkeeperClientSecondaryIsolationGroups" item is for this purpose.

**Comment by XuanYang-cn** [comment]
### BackGround

 Facts:

- helm install Milvus cluster on EKS with 2 datanodes.
- Milvus 2.0.0rc2
- Pulsar standalone

SDK operation:

- Perform insert on a collection with 100k batches of 128-dimensional vectors.

Phenomenon:

- Unable to insert more than 5 million vectors (2.5 GB), pulsar will refuse service due to insufficient storage space.
- Pulsar's disk usage has not dropped.

Issue7015 [^1] and issue7013 [^2] raise these questions:

- Why is there insufficient storage space in Pulsar?
- Why does inserting 2.5 G data cause insufficient storage space in Pulsar?

### Q&A

#### 1. Why is there insufficient storage space in Pulsar?

First of all, I discovered that we have set a message retention policy[^4] for pulsar, and set it to 7 days. Therefore, it is suspected that the pulsar disk space is too high because the message will be retained for 7 days. 

But then I discovered that **our settings for retention policy were invalid**[^5], which means that Pulsar processes messages as the default behavior, that is, does not retain any messages that have been ACKed.

Since Pulsar does not retain ACKed messages, Pulsar's disk space should've decreased.

**[Q1] Why hasn’t the disk space of Pulsar decreased?**[^2]

[A1] After thoroughly investigating Pulsar’s message life cycle and what're Pulsar’s retention policy strategies designed to solve[^3][^6][^7]。

Only ACKed messages will be marked as deleteable by Pulsar, and the time for real deletion is also affected by many other factors. So messages marked as deleteable will be deleted from the disk after "**a period of time**" [^12][^13][^14].

**[Q2] Is it because of concurrent inserts, the production rate of DML Topic is much higher than the consumption rate.**

So a large number of UnACK messages accumulate in the backlog of Pulsar. These unACK messages can't be marked as deleteable, and eventually cause short-term UnACK messages to quickly take up a lot of disk space?

[A2] No. The experiment shows that [^8][^9], even if msgRateIn[^10] is increased to 50 msg/s[^note], msgRateOut[^10] on the consumer side is not lagging behind. The backlog remains almost empty.

[^note]: Insert with 10 processes , with 1000  batches of 128-dim vectors in one process, continue to insert until I  confirm the rates.

**[Q3] Is it because "a period of time" is too long, causing the ACKed messages on the disk accumulated, so that the disk is never released, which consumes a lot of disk space?**

[A3] Yes. "A period of time" is at least 4 hours / 50000 messages [^note2] plus some other facts. In the actual test, after 1 full day the acked messages are still there. (Under Retention default configuration)

[^note2]: 5M rows (2.5GB) will be divided into 250 messages (each 10MB), plus rootcoord keeps sending timestamp, after about 3.77 h, the boundary of 50,000 messages will be reached.

#### 2. Why 2.5GB data will cause insufficient storage space in Pulsar

First of all, helm installed Pulsar will map the data volume out[^3], and set a 5GB disk space size limit[^3]

Secondly, when 2.5GB data is written to disk by bookkeeper, bookkeeper will write a Journal first, named timestamp.txn [^11]. Then it'll write the real data to disk, named 0.log. That is, the data need twice the storage of itself  in the bookkeeper.

```shell
# Insert 256M data
root@abcasfasd:/pulsar/data/standalone/bookkeeper0$ du -ah
262M	./17ba6104e65.txn
4.0K	./VERSION
0		./1.log
4.0K	./lastId
260M	./0.log
0		./locations/LOCK
4.5M	./locations/000003.log
4.0K	./locations/CURRENT
128K	./locations/LOG
4.0K	./locations/MANIFEST-000001
8.0K	./locations/OPTIONS-000005
4.0K	./locations/IDENTITY
4.6M	./locations
0		./ledgers/LOCK
71M		./ledgers/000003.log
4.0K	./ledgers/CURRENT
128K	./ledgers/LOG
4.0K	./ledgers/MANIFEST-000001
8.0K	./ledgers/OPTIONS-000005
4.0K	./ledgers/IDENTITY
71M		./ledgers
4.0K	./lastMark
597M	.
```



Finally, the bookkeeper and puslar used by pulsar standalone are in the same docker container and are also limited by 5G disk space.

Therefore, 2.5GB data will fill up the 5G space limited by Pulsar.

### Solutions

1. Ensure that the bookkeeper disk space is more than 2 times the amount of data
2. When helm deploys pulsar standalone, config the volume space limit to a reasonable value (the default 5GB is not suitable for large data sets).
3. Don't use pulsar standalone, use pulsar cluster to separate bookkeeper from pulsar.
4. other

### References

[^1]: [Issue #7015](https://github.com/milvus-io/milvus/issues/7015)
[^2]: [Issue #7013](https://github.com/milvus-io/milvus/issues/7013)
[^3]:  [Pulsar standalone docker config](https://github.com/milvus-io/milvus-helm/blob/67e938ad3ba60e23ff5246bf0ad49d7189debb4e/charts/milvus/values.yaml#L320)
[^4]: [Pulsar Retention Policy](https://pulsar.apache.org/docs/en/cookbooks-retention-expiry)
[^5]: [Wrong config for pulsar retention, fixed on #7385](https://github.com/milvus-io/milvus/blob/5fe96b9e198ecc0904e1dda1912b4b676f608078/deployments/docker/dev/docker-compose.yml#L18)
[^6]: [Understanding Pulsar Message TTL, Backlog, and Retention. *Ming Luo*](https://medium.com/kafkaesqueio/understanding-pulsar-message-ttl-backlog-and-retention-2c38ddecf868)
[^7]: [Pulsar concepts message-retention-and-expiry](https://pulsar.apache.org/docs/en/2.7.3/concepts-messaging/#message-retention-and-expiry)
[^8]: [Pulsar-admin topics stats](http://pulsar.apache.org/docs/en/pulsar-admin/#stats-1)
[^9]: [Pulsar-admin topics stats-internal](http://pulsar.apache.org/docs/en/pulsar-admin/#stats-internal)
[^10]: [Pulsar administration-stats](https://pulsar.apache.org/docs/en/2.7.3/administration-stats/)
[^11]: [Bookkeeper journal](http://bookkeeper.apache.org/archives/docs/r4.4.0/bookkeeperInternals.html)
[^12]: [Pulsar: delete messages from namespaces](https://pulsar.apache.org/docs/en/cookbooks-retention-expiry/#delete-messages-from-namespaces)
[^13]: [Bookkeeper entry log size and rollover period](https://bookkeeper.apache.org/docs/4.11.1/reference/config/#entry-log-settings)
[^14]: [Bookkeeper garbage collection interval](https://bookkeeper.apache.org/docs/4.11.1/reference/config/#entry-log-compaction-settings)
